utils/zerovec/src/varzerovec/lengthless.rs

```diff
@@ -0,0 +1,132 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::components::VarZeroVecComponents;
+use super::*;
+use crate::ule::*;
+use core::marker::PhantomData;
+use core::mem;
+
+/// A slice representing the index and data tables of a VarZeroVec,
+/// *without* any length fields. The length field is expected to be stored elsewhere.
+///
+/// Without knowing the length this is of course unsafe to use directly.
+#[repr(transparent)]
+#[derive(PartialEq, Eq)]
+pub(crate) struct VarZeroLengthlessSlice<T: ?Sized, F = Index16> {
+    marker: PhantomData<(F, T)>,
+    /// The original slice this was constructed from
+    // Safety invariant: This field must have successfully passed through
+    // VarZeroVecComponents::parse_byte_slice_with_length() with the length
+    // associated with this value.
+    entire_slice: [u8],
+}
+
+impl<T: VarULE + ?Sized, F: VarZeroVecFormat> VarZeroLengthlessSlice<T, F> {
+    /// Obtain a [`VarZeroVecComponents`] borrowing from the internal buffer
+    ///
+    /// Safety: `len` must be the length associated with this value
+    #[inline]
+    pub(crate) unsafe fn as_components<'a>(&'a self, len: u32) -> VarZeroVecComponents<'a, T, F> {
+        unsafe {
+            // safety: VarZeroSlice is guaranteed to parse here
+            VarZeroVecComponents::from_bytes_unchecked_with_length(len, &self.entire_slice)
+        }
+    }
+
+    /// Parse a VarZeroLengthlessSlice from a slice of the appropriate format
+    ///
+    /// Slices of the right format can be obtained via [`VarZeroSlice::as_bytes()`]
+    pub fn parse_byte_slice<'a>(len: u32, slice: &'a [u8]) -> Result<&'a Self, UleError> {
+        let _ = VarZeroVecComponents::<T, F>::parse_byte_slice_with_length(len, slice)
+            .map_err(|_| UleError::parse::<Self>())?;
+        unsafe {
+            // Safety: We just verified that it is of the correct format.
+            Ok(Self::from_bytes_unchecked(slice))
+        }
+    }
+
+    /// Uses a `&[u8]` buffer as a `VarZeroLengthlessSlice<T>` without any verification.
+    ///
+    /// # Safety
+    ///
+    /// `bytes` need to be an output from [`VarZeroLengthlessSlice::as_bytes()`], or alternatively
+    /// be valid to be passed to `from_bytes_unchecked_with_length`
+    ///
+    /// The length associated with this value will be the length associated with the original slice.
```

Comment:
Suggestion: for these functions I usually like to say that the safety invariant is that the slice must be parseable via `parse_byte_slice`; less action-at-a-distance.